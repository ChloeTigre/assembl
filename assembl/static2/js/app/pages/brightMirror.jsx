import React from 'react';
import { compose, graphql } from 'react-apollo';
import ThematicsQuery from '../graphql/ThematicsQuery.graphql';
import manageErrorAndLoading from '../components/common/manageErrorAndLoading';
import { browserHistory } from '../router';
import { get } from '../utils/routeMap';
import { displayAlert } from '../utils/utilityManager';

class BrightMirror extends React.Component {
  componentDidMount() {
    const { thematics, params } = this.props;
    const isParentRoute = !params.themeId;
    // After fetching thematics, redirect to the only BM thematic
    if (isParentRoute && thematics.length > 0) {
      const themeId = thematics[0].id;
      browserHistory.push(`${get('idea', { slug: params.slug, phase: params.phase, themeId: themeId })}`);
    }
  }

  render() {
    const { errors, identifier, children, params, phaseId } = this.props;
    if (errors) {
      displayAlert('danger', `${errors}`);
      return <div />;
    }
    const childrenElm = React.Children.map(children, child =>
      React.cloneElement(child, {
        id: params.themeId,
        identifier: identifier,
        phaseId: phaseId
      })
    );
    return <div className="bright-mirror">{childrenElm}</div>;
  }
}

export default compose(
  graphql(ThematicsQuery, {
    props: ({ data }) => {
      if (data.error || data.loading) {
        return {
          error: data.error,
          loading: data.loading
        };
      }

      return {
        error: data.error,
        loading: data.loading,
        thematics: data.thematics
      };
    }
  }),
  manageErrorAndLoading({ displayLoader: true })
)(BrightMirror);