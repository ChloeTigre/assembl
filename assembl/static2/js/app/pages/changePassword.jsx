/* eslint no-alert: "off" */

import React from 'react';
import { Grid, Col, FormGroup, FormControl, Button } from 'react-bootstrap';
import { Translate, I18n } from 'react-redux-i18n';
import { Link } from 'react-router';
import { getAuthorizationToken, getDiscussionSlug } from '../utils/globalFunctions';
import { postChangePassword } from '../services/authenticationService';
import inputHandler from '../utils/inputHandler';
import { get, getMaybeContextual } from '../utils/routeMap';
import { displayAlert } from '../utils/utilityManager';

class ChangePassword extends React.Component {
  constructor(props) {
    super(props);
    this.handleChangePassword = this.handleChangePassword.bind(this);
    this.submitForm = this.submitForm.bind(this);

    this.state = {
      token: getAuthorizationToken(this.props.location),
      disableButton: false
    };
  }

  handleChangePassword(e) {
    inputHandler(this, e);
  }

  submitForm(e) {
    e.preventDefault();
    this.setState({ disableButton: true });
    const payload = this.state;
    postChangePassword(payload)
      .then(() => {
        const slug = getDiscussionSlug();
        let route;
        let url;
        if (slug) {
          route = get('home', { slug: slug });
          url = new URL(route, window.location.origin);
          window.location = url;
        } else {
          route = get('root');
          url = new URL(route, window.location.origin);
          window.location = url;
        }
      })
      .catch((error) => {
        let msg;
        if (error instanceof Error) {
          if (error.name === 'PasswordMismatchError') {
            msg = I18n.t('login.incorrectPassword');
            this.setState({ disableButton: false });
            displayAlert('danger', msg, true);
          }
        } else {
          try {
            const firstError = error[0];
            displayAlert('danger', firstError.message, true);
          } catch (exception) {
            msg = I18n.t('login.somethingWentWrong');
            displayAlert('danger', msg, true);
          }
        }
      });
  }

  render() {
    const { disableButton } = this.state;
    return (
      <Grid fluid className="login-grid">
        <Col xs={12} md={6} className="login-container col-centered center">
          <div className="box-title">
            <Translate value="login.changePassword" />
          </div>
          <div className="box">
            <form>
              <div className="warning-label">
                <span className="warning">
                  <Translate value="login.oldPasswordWarning" />
                </span>
              </div>
              <FormGroup className="margin-m">
                <FormControl
                  type="password"
                  name="password1"
                  required
                  placeholder={I18n.t('login.newPassword')}
                  onChange={this.handleChangePassword}
                />
              </FormGroup>
              <FormGroup className="margin-m">
                <FormControl
                  type="password"
                  name="password2"
                  required
                  placeholder={I18n.t('login.newPassword2')}
                  onChange={this.handleChangePassword}
                />
              </FormGroup>
              <FormGroup className="margin-l">
                <Button
                  type="submit"
                  name="change_password"
                  value={I18n.t('login.changePassword')}
                  className="button-submit button-dark"
                  onClick={this.submitForm}
                  disabled={disableButton}
                >
                  <Translate value="login.changePassword" />
                </Button>
              </FormGroup>
              <FormControl.Static>
                <Translate
                  value="login.reRequestToken"
                  here={
                    <Link to={getMaybeContextual('requestPasswordChange')}>
                      <Translate value="here" />
                    </Link>
                  }
                />
              </FormControl.Static>
            </form>
          </div>
        </Col>
      </Grid>
    );
  }
}

export default ChangePassword;