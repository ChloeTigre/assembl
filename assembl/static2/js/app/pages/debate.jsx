// @flow
import * as React from 'react';
import { connect } from 'react-redux';
import { graphql } from 'react-apollo';
import Loader from '../components/common/loader';
import Themes from '../components/debate/common/themes';
import DebateThematicsQuery from '../graphql/DebateThematicsQuery.graphql';

type Props = {
  phaseId: string,
  identifier: string,
  data: {
    loading: boolean,
    error: ?Error,
    thematics: DebateThematicsQueryQuery
  },
  params: {
    phaseId?: string,
    themeId?: string,
    questionId?: string
  },
  children: React.Node
};

const Debate = ({ phaseId, identifier, data, params, children }: Props) => {
  const { loading, thematics } = data;
  const themeId = params.themeId || null;
  const questionId = params.questionId || null;
  const isParentRoute = !(themeId || questionId) || false;
  const childrenElm = React.Children.map(children, child =>
    React.cloneElement(child, {
      id: themeId || questionId,
      identifier: identifier,
      phaseId: phaseId
    })
  );
  return (
    <div className="debate">
      {loading && isParentRoute && <Loader color="black" />}
      <div>
        {thematics && isParentRoute && <Themes thematics={thematics} identifier={identifier} phaseId={phaseId} />}
        {thematics && !isParentRoute && <section className="debate-section">{childrenElm}</section>}
      </div>
    </div>
  );
};

const DebateWithData = graphql(DebateThematicsQuery)(Debate);

const mapStateToProps = state => ({
  lang: state.i18n.locale
});

export default connect(mapStateToProps)(DebateWithData);