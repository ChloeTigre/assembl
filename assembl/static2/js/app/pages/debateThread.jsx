// @flow
import * as React from 'react';
import { connect } from 'react-redux';
import { compose, graphql } from 'react-apollo';
import Loader from '../components/common/loader';
import Ideas from '../components/debate/tableOfIdeas/ideas';
import AllIdeasQuery from '../graphql/AllIdeasQuery.graphql';

type Props = {
  phaseId: string,
  identifier: string,
  data: {
    loading: boolean,
    error: ?Error,
    rootIdea: ?Idea, // rootIdea may be null if phase table of thematics wasn't saved
    ideas: Array<Idea>
  },
  params: {
    phaseId: string,
    themeId?: string,
    questionId?: string
  },
  children: React.Node
};

const DebateThread = ({ phaseId, identifier, data, params, children }: Props) => {
  if (!data) {
    return (
      <div className="debate">
        <Loader color="black" />
      </div>
    );
  }
  const { loading, ideas, rootIdea } = data;
  const themeId = params.themeId || null;
  const questionId = params.questionId || null;
  const isParentRoute = !(themeId || questionId) || false;
  const childrenElm = React.Children.map(children, child =>
    React.cloneElement(child, {
      id: themeId || questionId,
      identifier: identifier,
      phaseId: phaseId
    })
  );

  return (
    <div className="debate">
      {loading && isParentRoute && <Loader color="black" />}
      {!loading && ideas && isParentRoute && rootIdea ? (
        <Ideas key={phaseId} ideas={ideas} rootIdeaId={rootIdea.id} identifier={identifier} phaseId={phaseId} />
      ) : null}
      {!isParentRoute && <section className="debate-section">{childrenElm}</section>}
    </div>
  );
};

DebateThread.defaultProps = {
  data: null
};

const mapStateToProps = state => ({
  lang: state.i18n.locale,
  slug: state.debate.debateData.slug
});

export default compose(
  connect(mapStateToProps),
  graphql(AllIdeasQuery, {
    skip: ({ discussionPhaseId }) => !discussionPhaseId
  })
)(DebateThread);