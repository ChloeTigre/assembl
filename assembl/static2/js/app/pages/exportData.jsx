// @flow
import React from 'react';
import { compose, graphql } from 'react-apollo';
import moment from 'moment';
import { connect } from 'react-redux';
import ExportSection from '../components/administration/exportSection';
import { get } from '../utils/routeMap';
import { getDiscussionId } from '../utils/globalFunctions';
import manageErrorAndLoading from '../components/common/manageErrorAndLoading';

import DiscussionPreferences from '../graphql/DiscussionPreferences.graphql';

type Props = {
  section: string,
  locale: string,
  languages: Array<Language>,
  phases: Timeline
};

type State = {
  shouldTranslate: boolean,
  exportLocale: string,
  shouldBeAnonymous: boolean,
  start: ?moment,
  end: ?moment
};

class ExportData extends React.Component<Props, State> {
  state = {
    exportLocale: '',
    shouldTranslate: false,
    shouldBeAnonymous: false,
    start: null,
    end: null
  };

  handleDatesChange = ({ startDate, endDate }: DateRange) => this.setState({ start: startDate, end: endDate });

  handleShouldTranslate = (shouldTranslate: boolean) => {
    this.setState({ shouldTranslate: shouldTranslate });
  };

  toggleAnonymousOption = () => {
    this.setState(prevState => ({ shouldBeAnonymous: !prevState.shouldBeAnonymous }));
  };

  handleExportLocaleChange = (exportLocale: string) => {
    this.setState({
      exportLocale: exportLocale
    });
  };

  render() {
    const { section, languages, phases } = this.props;
    const { exportLocale, shouldBeAnonymous, shouldTranslate, start, end } = this.state;
    const locale = exportLocale || (languages && languages[0].locale);
    const translation = shouldTranslate && locale ? locale : '';
    const anonymous = `${shouldBeAnonymous.toString()}`;
    const debateId = getDiscussionId();
    const startDate = start ? start.format('L') : '';
    const endDate = end ? end.format('L') : '';
    const exportDataLink = get(
      'exportDebateData',
      { debateId: debateId },
      { translation: translation, anonymous: anonymous, startDate: startDate, endDate: endDate }
    );
    const exportTaxonomiesLink = get('exportTaxonomiesData', { debateId: debateId });
    return (
      <div>
        {section === '1' && (
          <ExportSection
            annotation="contributions"
            sectionTitle="contributions"
            handleDatesChange={this.handleDatesChange}
            handleAnonymousChange={this.toggleAnonymousOption}
            handleShouldTranslate={this.handleShouldTranslate}
            handleExportLocaleChange={this.handleExportLocaleChange}
            locale={locale}
            shouldBeAnonymous={shouldBeAnonymous}
            shouldTranslate={shouldTranslate}
            exportLocale={exportLocale}
            exportLink={exportDataLink}
            phases={phases}
            start={start}
            end={end}
            languages={languages}
          />
        )}
        {section === '2' && (
          <ExportSection sectionTitle="taxonomySectionTitle" annotation="taxonomyAnnotation" exportLink={exportTaxonomiesLink} />
        )}
      </div>
    );
  }
}

const mapStateToProps = ({ timeline, i18n }) => ({
  phases: timeline,
  locale: i18n.locale
});

export default compose(
  connect(mapStateToProps),
  graphql(DiscussionPreferences, {
    options: ({ locale }) => ({
      variables: {
        inLocale: locale
      }
    }),
    props: ({ data }) => {
      if (data.error || data.loading) {
        return {
          error: data.error,
          loading: data.loading
        };
      }

      return {
        error: data.error,
        loading: data.loading,
        languages: data.discussionPreferences.languages
      };
    }
  }),
  manageErrorAndLoading({ displayLoader: true })
)(ExportData);