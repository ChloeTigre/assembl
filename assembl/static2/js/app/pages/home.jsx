import React from 'react';
import { I18n } from 'react-redux-i18n';
import { connect } from 'react-redux';
import debounce from 'lodash/debounce';

import Header from '../components/home/header';
import Objectives from '../components/home/objectives';
import Phases from '../components/home/phases';
import Video from '../components/home/video';
import Twitter from '../components/home/twitter';
import Chatbot from '../components/home/chatbot';
import Partners from '../components/home/partners';
import ScrollOnePageButton from '../components/common/scrollOnePageButton';
import MessagePage from '../components/common/messagePage';

class Home extends React.Component {
  state = {
    scrollOnePageButtonHidden: false
  };

  componentWillMount() {
    window.addEventListener('scroll', this.hideScrollOnePageButton);
  }

  componentWillUnmount() {
    window.removeEventListener('scroll', this.hideScrollOnePageButton);
  }

  hideScrollOnePageButton = debounce(() => {
    if (!this.state.scrollOnePageButtonHidden) {
      this.setState(() => ({
        scrollOnePageButtonHidden: true
      }));
    }
  }, 100);

  render() {
    const { debateData } = this.props.debate;
    const { locale } = this.props.i18n;
    const { timeline } = this.props;
    if (!timeline) {
      // timeline is still loading
      return null;
    }
    if (timeline.length === 0) {
      return <MessagePage title={I18n.t('home.assemblNotConfigured')} text={I18n.t('administration.noTimeline')} />;
    }

    return (
      <div className="home">
        <Header />
        <ScrollOnePageButton hidden={this.state.scrollOnePageButtonHidden} />
        {debateData.objectives && <Objectives />}
        {timeline && <Phases />}
        {debateData.video && <Video />}
        {debateData.twitter && <Twitter />}
        {debateData.chatbot && <Chatbot chatbot={debateData.chatbot} locale={locale} />}
        {debateData.partners && <Partners />}
      </div>
    );
  }
}

const mapStateToProps = state => ({
  debate: state.debate,
  timeline: state.timeline,
  i18n: state.i18n
});

export default connect(mapStateToProps)(Home);