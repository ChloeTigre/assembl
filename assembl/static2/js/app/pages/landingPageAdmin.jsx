// @flow
import React from 'react';
import { compose, graphql } from 'react-apollo';
import { connect } from 'react-redux';
import { type Route, type Router } from 'react-router';
import { I18n } from 'react-redux-i18n';
import { EditorState } from 'draft-js';
import { List, Map } from 'immutable';
import moment from 'moment';

import { convertEntriesToHTML, convertImmutableEntriesToJS } from '../utils/draftjs';
import { getEntryValueForLocale } from '../utils/i18n';
import ManageModules from '../components/administration/landingPage/manageModules';
import CustomizeHeader from '../components/administration/landingPage/customizeHeader';
import ManageTimeline from '../components/administration/landingPage/manageTimeline';
import Navbar from '../components/administration/navbar';
import { displayAlert } from '../utils/utilityManager';
import SaveButton, { getMutationsPromises, runSerial } from '../components/administration/saveButton';
import updateDiscussionMutation from '../graphql/mutations/updateDiscussion.graphql';
import updateDiscussionPhaseMutation from '../graphql/mutations/updateDiscussionPhase.graphql';
import createLandingPageModule from '../graphql/mutations/createLandingPageModule.graphql';
import updateLandingPageModule from '../graphql/mutations/updateLandingPageModule.graphql';

type Props = {
  createLandingPageModule: Function,
  updateLandingPageModule: Function,
  landingPageModules: Array<Object>,
  landingPageModulesHasChanged: boolean,
  refetchLandingPageModules: Function,
  refetchLandingPage: Function,
  route: Route,
  router: Router,
  section: string,
  editLocale: string,
  header: {
    title: string,
    subtitle: EditorState,
    buttonLabel: string,
    headerImgMimeType: string,
    headerImgUrl: string,
    headerImgTitle: string,
    logoImgMimeType: string,
    logoImgUrl: string,
    logoImgTitle: string
  },
  pageHasChanged: boolean,
  phasesHaveChanged: boolean,
  page: Map<string, any>,
  updateDiscussion: Function,
  updateDiscussionPhase: Function,
  discussionPhases: Array<Object>,
  refetchTimeline: Function,
  timelineModuleId: string
};

type State = {
  refetching: boolean
};

class LandingPageAdmin extends React.Component<Props, State> {
  constructor() {
    super();
    this.state = {
      refetching: false
    };
  }

  componentDidMount() {
    this.props.router.setRouteLeaveHook(this.props.route, this.routerWillLeave);
  }

  componentWillUnmount() {
    this.props.router.setRouteLeaveHook(this.props.route, null);
  }

  routerWillLeave = () => {
    if (this.dataHaveChanged() && !this.state.refetching) {
      return I18n.t('administration.confirmUnsavedChanges');
    }

    return null;
  };

  getImageVariable = (img) => {
    const externalUrl = img ? img.externalUrl : null;
    if (externalUrl === 'TO_DELETE') {
      return externalUrl;
    }
    return typeof externalUrl === 'object' ? externalUrl : null;
  };

  createVariablesForDiscussionPhaseMutation = phase => ({
    id: phase.id,
    identifier: phase.identifier,
    order: phase.order,
    isThematicsTable: phase.isThematicsTable,
    start: moment(phase.start, moment.ISO_8601),
    end: moment(phase.end, moment.ISO_8601),
    titleEntries: phase.titleEntries,
    descriptionEntries: phase.descriptionEntries,
    image: this.getImageVariable(phase.image)
  });

  createVariablesForLandingPageModuleMutation = item => ({
    configuration: '{}',
    enabled: item.enabled,
    order: item.order,
    typeIdentifier: item.moduleType.identifier,
    titleEntries: item.titleEntries,
    subtitleEntries: item.subtitleEntries
  });

  saveAction = () => {
    const {
      landingPageModulesHasChanged,
      landingPageModules,
      refetchLandingPageModules,
      refetchLandingPage,
      pageHasChanged,
      page,
      updateDiscussion,
      phasesHaveChanged,
      discussionPhases,
      editLocale,
      updateDiscussionPhase,
      refetchTimeline
    } = this.props;
    displayAlert('success', `${I18n.t('loading.wait')}...`, false, -1);
    if (landingPageModulesHasChanged) {
      const mutationsPromises = getMutationsPromises({
        items: landingPageModules,
        variablesCreator: this.createVariablesForLandingPageModuleMutation,
        createMutation: this.props.createLandingPageModule,
        updateMutation: this.props.updateLandingPageModule
      });

      runSerial(mutationsPromises)
        .then(() => {
          refetchLandingPageModules().then(() => this.setState({ refetching: false }));
          displayAlert('success', I18n.t('administration.landingPage.successSave'));
        })
        .catch((error) => {
          displayAlert('danger', error, false, 30000);
        });
    }

    if (pageHasChanged) {
      // $FlowFixMe flow doesn't seem to know the second param of Map.get()
      const subtitleEntries = convertImmutableEntriesToJS(page.get('subtitleEntries', List()));
      updateDiscussion({
        variables: {
          // $FlowFixMe flow doesn't seem to know the second param of Map.get()
          titleEntries: page.get('titleEntries', List()).toJS(),
          subtitleEntries: convertEntriesToHTML(subtitleEntries),
          // $FlowFixMe flow doesn't seem to know the second param of Map.get()
          buttonLabelEntries: page.get('buttonLabelEntries', List()).toJS(),
          // $FlowFixMe flow doesn't seem to know the second param of Map.get()
          headerImage: this.getImageVariable(page.get('headerImage', Map()).toJS()),
          // $FlowFixMe flow doesn't seem to know the second param of Map.get()
          logoImage: this.getImageVariable(page.get('logoImage', Map()).toJS())
        }
      })
        .then(() => {
          refetchLandingPage().then(() => this.setState({ refetching: false }));
          displayAlert('success', I18n.t('administration.landingPage.headerSuccessSave'));
        })
        .catch((error) => {
          displayAlert('danger', error.message);
        });
    }

    if (phasesHaveChanged) {
      const mutationPromises = getMutationsPromises({
        items: discussionPhases,
        variablesCreator: this.createVariablesForDiscussionPhaseMutation,
        updateMutation: updateDiscussionPhase,
        lang: editLocale
      });

      runSerial(mutationPromises)
        .then(() => {
          this.setState({ refetching: true }, () => {
            refetchTimeline().then(() => {
              displayAlert('success', I18n.t('administration.timelineAdmin.successSave'));
              this.setState({ refetching: false });
            });
          });
        })
        .catch((error) => {
          displayAlert('danger', `${error}`, false, 30000);
        });
    }
  };

  dataHaveChanged = (): boolean =>
    this.props.landingPageModulesHasChanged || this.props.pageHasChanged || this.props.phasesHaveChanged;

  render() {
    const { editLocale, header, section, timelineModuleId } = this.props;
    const saveDisabled = !this.dataHaveChanged();
    return (
      <div className="landing-page-admin">
        <SaveButton disabled={saveDisabled} saveAction={this.saveAction} />
        {section === '1' && <ManageModules {...this.props} />}
        {section === '2' && <CustomizeHeader editLocale={editLocale} header={header} />}
        {section === '3' && <ManageTimeline timelineModuleId={timelineModuleId} editLocale={editLocale} />}
        {section && <Navbar currentStep={section} steps={['1', '2', '3']} phaseIdentifier="landingPage" />}
      </div>
    );
  }
}

const mapStateToProps = ({ admin: { editLocale, landingPage, timeline } }) => {
  const { page } = landingPage;
  const subtitle = getEntryValueForLocale(page.get('subtitleEntries'), editLocale, EditorState.createEmpty());
  const timelineModule = landingPage.modulesById.find(module => module.getIn(['moduleType', 'identifier']) === 'TIMELINE');
  // timelineModule can be undefined when modulesById is not loaded yet
  const timelineModuleId = timelineModule ? timelineModule.get('id') : null;
  return {
    landingPageModulesHasChanged: landingPage.modulesHasChanged,
    landingPageModules: landingPage.modulesById
      .map((module) => {
        const id = module.get('id');
        const idx = landingPage.enabledModulesInOrder.indexOf(id);
        return module
          .set('order', idx + 1)
          .set('_hasChanged', true)
          .set('_isNew', !module.get('existsInDatabase'));
      })
      .valueSeq()
      .toJS(),
    timelineModuleId: timelineModuleId,
    header: {
      title: getEntryValueForLocale(page.get('titleEntries'), editLocale, ''),
      subtitle: subtitle,
      buttonLabel: getEntryValueForLocale(page.get('buttonLabelEntries'), editLocale, ''),
      headerImgMimeType: page.getIn(['headerImage', 'mimeType']),
      headerImgUrl: page.getIn(['headerImage', 'externalUrl']),
      headerImgTitle: page.getIn(['headerImage', 'title']),
      logoImgMimeType: page.getIn(['logoImage', 'mimeType']),
      logoImgUrl: page.getIn(['logoImage', 'externalUrl']),
      logoImgTitle: page.getIn(['logoImage', 'title'])
    },
    page: landingPage.page,
    pageHasChanged: landingPage.pageHasChanged,
    editLocale: editLocale,
    phasesHaveChanged: timeline.phasesHaveChanged,
    discussionPhases: timeline.phasesById
      .sortBy(phase => phase.get('order'))
      .valueSeq()
      .toJS()
  };
};

export default compose(
  connect(mapStateToProps),
  graphql(createLandingPageModule, {
    name: 'createLandingPageModule'
  }),
  graphql(updateLandingPageModule, {
    name: 'updateLandingPageModule'
  }),
  graphql(updateDiscussionMutation, {
    name: 'updateDiscussion'
  }),
  graphql(updateDiscussionPhaseMutation, {
    name: 'updateDiscussionPhase'
  })
)(LandingPageAdmin);