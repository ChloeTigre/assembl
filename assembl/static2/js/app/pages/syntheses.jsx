// @flow
import React from 'react';
import { connect } from 'react-redux';
import { compose, graphql } from 'react-apollo';
import { Link } from 'react-router';
import { I18n, Translate } from 'react-redux-i18n';

import { get } from '../utils/routeMap';
import Section from '../components/common/section';
import SynthesesList from '../components/synthesis/synthesesList';
import SynthesesQuery from '../graphql/SynthesesQuery.graphql';
import manageErrorAndLoading from '../components/common/manageErrorAndLoading';
import { DELETE_CALLBACK } from '../constants';
import { displayAlert } from '../utils/utilityManager';
import type { SynthesisPost } from '../types.flow';

type SynthesesProps = {
  syntheses: Array<SynthesisPost>,
  slug: string,
  hasSyntheses: boolean,
  lang: string,
  location: { state: { callback: string } }
};

export class DumbSyntheses extends React.Component<SynthesesProps> {
  componentDidMount() {
    this.displayDeleteMessage();
  }

  displayDeleteMessage() {
    // Location state is set in brightMirrorFiction.jsx > deleteFictionCallback
    const locationState = this.props.location.state;
    if (locationState && locationState.callback === DELETE_CALLBACK) {
      displayAlert('success', I18n.t('debate.syntheses.deleteSuccessMessage'));
    }
  }

  render() {
    const { syntheses, slug, hasSyntheses, lang } = this.props;
    const createSynthesisRoute = get('createSynthesis', { slug: slug });
    const updateSynthesesQuery = {
      query: SynthesesQuery,
      variables: {
        lang: lang
      }
    };
    const refetchQueries = [updateSynthesesQuery];

    return (
      <Section title="debate.syntheses.summary" translate>
        <div className="center create-synthesis-button">
          <Link to={createSynthesisRoute} className="button-submit button-dark">
            <Translate value="debate.syntheses.createNewSynthesis" />
          </Link>
        </div>
        {!hasSyntheses ? (
          <h2 className="dark-title-2 margin-left-xxl">
            <Translate value="synthesis.noSynthesisYet" />
          </h2>
        ) : (
          <SynthesesList syntheses={syntheses} refetchQueries={refetchQueries} />
        )}
      </Section>
    );
  }
}

const mapStateToProps = state => ({
  lang: state.i18n.locale,
  slug: state.debate.debateData.slug
});

export default compose(
  connect(mapStateToProps),
  graphql(SynthesesQuery, {
    props: ({ data }) => {
      if (data.error || data.loading) {
        return {
          error: data.error,
          loading: data.loading
        };
      }

      return {
        error: data.error,
        loading: data.loading,
        syntheses: data.syntheses,
        hasSyntheses: data && data.syntheses ? data.syntheses.length > 0 : false
      };
    }
  }),
  manageErrorAndLoading({ displayLoader: true })
)(DumbSyntheses);