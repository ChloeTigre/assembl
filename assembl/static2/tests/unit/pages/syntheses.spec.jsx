import React from 'react';
import renderer from 'react-test-renderer';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16.3';

import { DumbSyntheses } from '../../../js/app/pages/syntheses';
import SynthesesList from '../../../js/app/components/synthesis/synthesesList';

configure({ adapter: new Adapter() });

describe('Syntheses component - with shallow', () => {
  let wrapper;
  let synthesesProps;
  beforeEach(() => {
    synthesesProps = {
      location: { state: undefined },
      hasSyntheses: true,
      syntheses: [
        {
          id: 'barId',
          subject: 'Bar',
          img: { externalUrl: 'http://foo.com/bar' },
          creationDate: '2017-02-10T09:15:20.707854+00:00',
          post: { id: 'someId' }
        },
        {
          id: 'fooId',
          subject: 'Foo',
          img: { externalUrl: 'http://foo.com/bar' },
          creationDate: '2018-02-10T09:15:20.707854+00:00',
          post: { id: 'someId' }
        }
      ],
      slug: 'fooslug'
    };
  });

  it('should render syntheses list', () => {
    wrapper = shallow(<DumbSyntheses {...synthesesProps} />);
    expect(wrapper.find('SynthesesList')).toHaveLength(1);
  });

  it('should render synthesis previews', () => {
    wrapper = shallow(<SynthesesList syntheses={synthesesProps.syntheses} />);
    expect(wrapper.find('SynthesisPreview')).toHaveLength(2);
  });

  it('should render synthesis previews in state order first then in date order', () => {
    wrapper = shallow(<SynthesesList syntheses={synthesesProps.syntheses} />);
    const listPreviews = wrapper.find('SynthesisPreview');
    expect(listPreviews.get(0).props.synthesis.subject).toEqual('Foo');
    expect(listPreviews.get(1).props.synthesis.subject).toEqual('Bar');
  });
});

describe('Syntheses component', () => {
  it('should match empty Syntheses snapshot', () => {
    const props = { location: { state: undefined }, hasSyntheses: false, syntheses: [] };
    const rendered = renderer.create(<DumbSyntheses {...props} />).toJSON();
    expect(rendered).toMatchSnapshot();
  });
});