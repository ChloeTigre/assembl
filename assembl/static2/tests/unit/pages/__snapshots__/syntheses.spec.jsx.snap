// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Syntheses component should match empty Syntheses snapshot 1`] = `
<div
  className="container-fluid"
>
  <div
    className="max-container syntheses-page"
  >
    <div
      className="title-section"
    >
      <div
        className="title-hyphen"
      />
      <h1
        className="dark-title-1 no-margin-left-xs"
      >
        <span>
          Syntheses summary
        </span>
      </h1>
    </div>
    <div
      className="page-body"
    >
      <div
        className="center margin-left-xxl syntheses-no-synthesis"
      >
        <span>
          There is no synthesis available yet.
        </span>
      </div>
    </div>
  </div>
</div>
`;
